### ISSUE TITLE
Unclear iteractions between `iterate()` methods and context/header parsing

### ISSUE THREAD

**Issue by djc** [uniVocity/univocity-parsers#314]
For my use case, I want to use the header of a CSV file to determine which parser to use:

1. Get header
2. Check header against all parsers to find the matching parser
3. Let matching parser handle parse the file

However, while trying to use your library I found it quite hard to figure out how I could access the header and get it to be set correctly in the `RecordMetadata`. I finally figured out that I should call `parsedHeaders()` on the `ParsingContext` to trigger reading of the header (but this method is not mentioned in the tutorial at all).

However, because `iterate()` and `iterateRecords()`, like `beginParsing()`, take the input reader as an argument, this makes it seem impossible to both get the headers and use the iterators (at least my tests to call `iter.getContext().parsedHeaders()` did not succeed -- I suppose getting the iterator does not imply `beginParsing()`, but `IterateResult` also does not expose a `beginParsing()` method I could use to trigger this.

This means I reverted to a simple while loop calling `parser.parseNextRecord()`, but this ends up being verbose because my parsing loop has a few cases of `continue`, meaning I have to repeat the `parser.parseNextRecord()` call in a number of places.

Maybe I'm missing something, but it seems like this situation could be improved. Whether that means some documentation improvements or I missed some particular APIs, I'm not sure.

**Comment by jbax** [comment]
Can you please provide some sample code I can run and test to address the problem you are having?

**Comment by djc** [comment]
So my current approach looks like this:

```kotlin
val settings = CsvParserSettings()
settings.detectFormatAutomatically()
settings.isHeaderExtractionEnabled = true
val parser = CsvParser(settings)
val iter = parser.iterateRecords(data)
iter.context.parsedHeaders() // this throws a NullPointerException

return listOf(GcFileParser(iter), IpcFileParser(iter), IspFileParser(iter)).first {
    it.matches()
}
```

Where `matches()` is like this:

```kotlin
fun matches(): Boolean {
    return listOf(amountColumnName, customerColumnName, groupColumnName).all {
        csv.context.recordMetaData().containsColumn(it)
    }
}
```

And then I iterate over the results like this:

```kotlin
for (row in csv) {
    if (!includeRow(row)) continue
    val rate = getRate(row) ?: continue
    val customer = getCustomer(row, receivers) ?: continue
}
```

**Comment by jbax** [comment]
Thanks, that was very helpful and I was able to reproduce the bug. 

Basically the error happened because the parser was not starting until you a call to `iter.iterator.hasNext` happened. This will be executed automatically if you use a for loop, avoiding the bug:

```
    for(Record record : parser.iterateRecords(data)){
        row.metaData().containsColumn("col"); //should work
    }
```

Anyway I've fixed this issue and released a 2.8.2-SNAPSHOT build which you can use already and should work for you without any code changes.

Thank you for using our parsers!

**Comment by djc** [comment]
Awesome! Please consider also adding some documentation to the tutorials that discuss how and when the headers can be extracted/accessed/parsed.
